OffscreenCanvas rendering can hide the border from the element.

mattwoodrow opened the pull request:
#### d71f4ec5052b4d57ddd2e046b22f0556cbce5704
<pre>
OffscreenCanvas rendering can hide the border from the element.
<a href="https://bugs.webkit.org/show_bug.cgi?id=254493">https://bugs.webkit.org/show_bug.cgi?id=254493</a>

Reviewed by Simon Fraser.

This initializes the &apos;contents layer&apos; of GraphicsLayerCA so that the canvas buffer is presented
as a separate layer on top of the primary layer containing the border.

In order to prevent the main thread from changing/removing the contents of this platform layer, it
changes the async display delegate to extend the normal display delegate, so that the output buffer
of OffscreenCanvas is provided to the layer both on the main-thread as well as asynchronously sending
it to the UI-side compositor.

Since we potentially have a race with this, where the main-thread RemoteLayerTransaction might send a buffer
that is a frame behind the one most recently provided asynchronously, it also adds tracking of the RenderingResourceIdentifiers,
and skips the main-thread buffer update if we have a newer one from the async path.

This also exposed another bug, where if we sent a buffer asynchronously for a layer that didn&apos;t have a backing
store (due to being offscreen), it would get lost. If the layer then moved onscreen, we&apos;d wouldn&apos;t have the contents (until
the page drew to the OffscreenCanvas again).

* LayoutTests/fast/canvas/offscreen-backing-store-attached-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-backing-store-attached.html: Added.

New test for the case where the canvas doesn&apos;t have a backing store at the time offscreen canvas sends a buffer, but has
a backing store later.

* LayoutTests/fast/canvas/offscreen-border-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-border.html: Added.

Test for offscreen canvas and a border on the same element.

* LayoutTests/fast/canvas/offscreen-clipped-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-clipped.html: Added.

Test for offscreen canvas and clip path, to ensure clipping of the &apos;contents layer&apos; works correctly.

* Source/WebCore/platform/graphics/GraphicsLayerContentsDisplayDelegate.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToAsyncDisplayDelegate):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::setContents):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::asyncContentsIdentifier const):
(WebKit::RemoteLayerTreeNode::setAsyncContentsIdentifier):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemoteAsyncContentsDisplayDelegate::GraphicsLayerCARemoteAsyncContentsDisplayDelegate):
(WebKit::GraphicsLayerCARemoteAsyncContentsDisplayDelegate::setDestinationLayerID):
(WebKit::GraphicsLayerCARemote::createAsyncContentsDisplayDelegate):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::setContents):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:

Canonical link: <a href="https://commits.webkit.org/263731@main">https://commits.webkit.org/263731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05b325ae138ae5259c85a0cd7dc3cc9c1e8b7feb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5856 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7066 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5529 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5518 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5892 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5641 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6995 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4939 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7101 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3158 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4956 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12094 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5024 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5036 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6789 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5472 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4484 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4908 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9031 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/640 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5289 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -1343,6 +1343,20 @@ void GraphicsLayerCA::setContentsDisplayDelegate(RefPtr<GraphicsLayerContentsDis
+    if (!m_contentsLayer) {
+        m_contentsLayer = createPlatformCALayer(PlatformCALayer::LayerTypeContentsProvidedLayer, this);
+        m_contentsLayerPurpose = purpose;
+        m_contentsLayer->setBackingStoreAttached(false);
```
It's a bit weird to call this here. `setBackingStoreAttached(false)` is normally about things outside the viewport.

mattwoodrow replied:
This is necessary (or some other equivalent change), since we're pushing the buffer asynchronously to the UI-side layers.
The default value is true, and with that we'll end up trying to allocate a buffer for the RemoteLayerBackingStore and painting it, which fails (and we don't want it to work).

This works for normal delegated contents layers (like DOM <canvas> + WebGL), since that has special code paths in RemoteLayerBackingStore which bypasses the buffer allocation, and calls the delegate instead of paint().

smfr replied:
normally `drawsContent` is what determines if we allocate backing store for a layer, and it’s an error to call `setNeedsDisplay` on a layer which has `drawsContent = false`

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -1343,6 +1343,20 @@ void GraphicsLayerCA::setContentsDisplayDelegate(RefPtr<GraphicsLayerContentsDis
+        m_contentsLayer = createPlatformCALayer(PlatformCALayer::LayerTypeContentsProvidedLayer, this);
+        m_contentsLayerPurpose = purpose;
+        m_contentsLayer->setBackingStoreAttached(false);
+        m_contentsLayer->setAcceleratesDrawing(true);
```
Also weird to force this on.

mattwoodrow replied:
This was just cargo-culted from the what we do for normal delegated display, I can remove it.

smfr commented on LayoutTests/fast/canvas/offscreen-backing-store-attached-expected.html:
```diff
@@ -0,0 +1,17 @@
+
+const square = new Path2D();
+square.rect(50, 50, 100, 100);
+context.fillStyle = 'red';
```
Don't use red in a passing test.

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
```diff
@@ -153,6 +153,7 @@ class GraphicsLayerCA : public GraphicsLayer, public PlatformCALayerClient {
     WEBCORE_EXPORT void setContentsToPlatformLayerHost(LayerHostingContextIdentifier) override;
     WEBCORE_EXPORT void setContentsToVideoElement(HTMLVideoElement&, ContentsLayerPurpose) override;
     WEBCORE_EXPORT void setContentsDisplayDelegate(RefPtr<GraphicsLayerContentsDisplayDelegate>&&, ContentsLayerPurpose) override;
+    WEBCORE_EXPORT PlatformLayerIdentifier setContentsToAsyncDisplayDelegate(RefPtr<GraphicsLayerContentsDisplayDelegate>, ContentsLayerPurpose);
```
Should the first arg be a `RefPtr<>&&` ?

mattwoodrow replied:
I don't think so, we want to hold a new reference to the delegate (since the caller wants to still have a reference too).

smfr commented on Source/WebCore/platform/graphics/ca/PlatformCALayerDelegatedContents.h:
```diff
@@ -50,6 +51,7 @@ struct PlatformCALayerDelegatedContents {
     using FinishedIdentifier = ObjectIdentifier<Detail::PlatformCALayerDelegatedContentsIdentifier>;
     MachSendRight surface;
     FinishedIdentifier finishedIdentifier;
+    std::optional<RenderingResourceIdentifier> surfaceIdentifier { std::nullopt };
```
No need for the initializer.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
```diff
@@ -30,6 +30,7 @@
 
 #import "ArgumentCoders.h"
 #import "GraphicsLayerCARemote.h"
+#import "Logging.h"
```
Unused or unified build fix?

mattwoodrow replied:
Unified build fix (as a result of converting GraphicsLayer.cpp to .mm I believe).

mattwoodrow committed:
OffscreenCanvas rendering can hide the border from the element.
https://bugs.webkit.org/show_bug.cgi?id=254493

Reviewed by Simon Fraser.

This initializes the 'contents layer' of GraphicsLayerCA so that the canvas buffer is presented
as a separate layer on top of the primary layer containing the border.

In order to prevent the main thread from changing/removing the contents of this platform layer, it
changes the async display delegate to extend the normal display delegate, so that the output buffer
of OffscreenCanvas is provided to the layer both on the main-thread as well as asynchronously sending
it to the UI-side compositor.

Since we potentially have a race with this, where the main-thread RemoteLayerTransaction might send a buffer
that is a frame behind the one most recently provided asynchronously, it also adds tracking of the RenderingResourceIdentifiers,
and skips the main-thread buffer update if we have a newer one from the async path.

This also exposed another bug, where if we sent a buffer asynchronously for a layer that didn't have a backing
store (due to being offscreen), it would get lost. If the layer then moved onscreen, we'd wouldn't have the contents (until
the page drew to the OffscreenCanvas again).

* LayoutTests/fast/canvas/offscreen-backing-store-attached-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-backing-store-attached.html: Added.

New test for the case where the canvas doesn't have a backing store at the time offscreen canvas sends a buffer, but has
a backing store later.

* LayoutTests/fast/canvas/offscreen-border-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-border.html: Added.

Test for offscreen canvas and a border on the same element.

* LayoutTests/fast/canvas/offscreen-clipped-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-clipped.html: Added.

Test for offscreen canvas and clip path, to ensure clipping of the 'contents layer' works correctly.

* Source/WebCore/platform/graphics/GraphicsLayerContentsDisplayDelegate.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToAsyncDisplayDelegate):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::setContents):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::asyncSetLayerContents):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::asyncContentsIdentifier const):
(WebKit::RemoteLayerTreeNode::setAsyncContentsIdentifier):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemoteAsyncContentsDisplayDelegate::GraphicsLayerCARemoteAsyncContentsDisplayDelegate):
(WebKit::GraphicsLayerCARemoteAsyncContentsDisplayDelegate::setDestinationLayerID):
(WebKit::GraphicsLayerCARemote::createAsyncContentsDisplayDelegate):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.cpp:
(WebKit::PlatformCALayerRemote::setContents):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:

Canonical link: https://commits.webkit.org/263731@main
